### ISSUE TITLE
Issue running Webmidi.js with Next.js

### ISSUE THREAD

**Issue by rajangdavis** [djipco/webmidi#133]
**Describe the bug and how ro reproduce it**
Importing the WebMidi v3.0.0 library in Next.js while using Server Side Rendering creates issues due to the rendering context.

**Environment:**
Specify the environment where you are witnessing the problem:
 - Mac OSX
 - Chrome Version 88.0.4324.192 (Official Build) (x86_64)
 - Node v15.9.0
 - webmidi ^3.0.0-alpha.2
 - next 7.5.4

**Additional context**
The issue stems from this part of the code:
```
// If we are inside Node.js, polyfill navigator.requestMIDIAccess() and import performance.now()
    if (this.isNode) {

      // Important: performance must be imported before jzz because jzz checks for its existence at
      // startup and falls back to something less precise if absent.
      global.performance = require("perf_hooks").performance;
      global.navigator = require("jzz");

    }
```

When Next is running in a server-side rendered context, `this.isNode` evaluates to true and creates an error that the `perf_hook` module cannot be resolved while compiling.

A temporary work around that I have found is that I can turn off server side rendering and the errors go away. For what I am prototyping, this should be sufficient, but it would be ideal if I could be able to tap into SSR without requiring certain aspects of the library that I might not use.

I tried performing a dynamic import using a React hook; this allows developers to dynamically load dependencies/components while in a browser context. This generates the same errors previously mentioned.

I'm going to see if I can defer the install of the `perf_hooks` until after the enable function is called. If that doesn't work, do you have any suggestions for me to try? I have a few ideas but open to whatever direction you think.

**Comment by rajangdavis** [comment]
So it looks like this is a non-trivial fix... By design, the WebMIDI class is constructed and does all the server vs browser checking before it can be enabled or disabled. 

Even if the JZZ and Performance modules could get side stepped, Next.js will complain that when it tries to resolve the navigator object because it will run it in a server side context.

Not sure how to update the docs, but I would make a point that this library will only work with React/Next with SSR off. Making it work will be outside of the scope of this library, so closing this issue.

**Comment by djipco** [comment]
The platform check and polyfilling does not have to be in the `WebMidi` object constructor. It could happen later when calling `enable()`. By doing so, I could add an option to `enable()` to disable any sort of polyfilling. Would that help?

**Comment by rajangdavis** [comment]
Hi @djipco,

Thanks for the quick response.

I had tried to defer the polyfilling by putting the related code in the `enable` function.

The problem is that in the WebMidi specs, there is an `afterEach` hook that disables the WedMidi object; the `disable` function will check for the `navigator` object and, if I recall correctly, there is a check for `isSupported` which also relies on the `navigator` object being present.

I could be mistaken, but if I think even if we bypass the `navigator` object, it will fail in a Server Side Rendering context because it is being ran in a Node context which lacks a `navigator`. 

I don't think this is really a problem with your library, I suspect I just haven't found the right pattern.

I *think* I can solve it using React Hooks such as [this example](https://stackoverflow.com/a/62991981) which should [technically only be run in a browser context](https://css-tricks.com/run-useeffect-only-once/).

I have something working that is similar to the above React Hooks pattern, but it's importing a flat JS file of the WebMidi.js library I downloaded from a CDN. Worst case scenario, _that_ might have to be the pattern for using the WebMidi.js library with React, but it's definitely better than working with the WebMIDI directly...

Will keep you updated with what I figure out.

Kind regards,
Raj

**Comment by djipco** [comment]
> The problem is that in the WebMidi specs, there is an afterEach hook that disables the WedMidi object; the disable function will check for the navigator object and, if I recall correctly, there is a check for isSupported which also relies on the navigator object being present.

You are right. I just changed that. There was no real need to throw an error in `disable()`. Version 3.0.0-alpha.3 includes the change. 

> I could be mistaken, but if I think even if we bypass the navigator object, it will fail in a Server Side Rendering context because it is being ran in a Node context which lacks a navigator.

When running with SSR enabled,  is the `jzz` module successfully included? If it is, it should now work, I think...

**Comment by rajangdavis** [comment]
My goal was to have SSR care only about the HTML and not the WebMIDI API and I was struggling to get that working.

I've ran into an issue with this on a different app with the Geolocation API; the strategy at that time was to use a React Lifecycle method to load the navigation object after the component mounted.

The newer preferable method is using React hooks to update state if it is reliant on the browser. I am still learning this newer abstraction, but I have something working. 

Further changes to the library are not required for what I need to do.

If anything, I will probably contribute some documentation for working with React/Next.js. 

I do think it might be a good idea to be able to opt in and out of features, it might be helpful for experimental features or for profiling performance.

**Comment by djipco** [comment]
> To be clear, further changes to the library are not required for what I need to do.

I understand. However, your issue got me to re-examine parts of the code and make improvements. It is very useful to get feedback from knowledgeable users.

> When I refactor that code into something "production ready", I'll make sure to contribute some documentation for working with React/Next.js.

Awesome. When you have something (and if you feel like it) you could add it to the `examples` folder of the `develop` branch. You could also post it to the [forum](https://webmidijs.org/forum/) so people know it exists even if it's for the upcoming version.

Thanks again for your feedback. 🙏

**Comment by djipco** [comment]
@rajangdavis I think I found a solution to this problem. You can read about it [here](https://github.com/djipco/webmidi/issues/137#issuecomment-819185495). If you had a second to try it out and let me know if it fixes the problem for you, it would be appreciated. Thanks.

**Comment by rajangdavis** [comment]
Hey @djipco, just wanted to let you know that I've seen your comment. 

Right now, I'm focusing on more on designing user interfaces for my MIDI project; after I get through that, I'll make sure to test out alpha-5. 

I have plans to update a lot of code and restructure a few things which would also include how I interact with the Webmidi.js library. Once I get to that, I can update the code no problem, I just have a few other things I'm prioritizing at the moment.

It sounds like it works for a few other users, so looking forward to it. I'm happy to hear that there is some progress happening in the library and that the community is able to move forward!

**Comment by rajangdavis** [comment]
@djipco Works like a charm! Thanks for the update!
